AK/BigIntBase.h

```diff
@@ -0,0 +1,566 @@
+        else if constexpr (operation == BitwiseOperation::INVERT)
+            result[i] = ~word1;
+        else
+            static_assert(((void)operation, false));
+    }
+}
+
+// See `storage_compute_bitwise` for the signedness requirements.
+template<BitwiseOperation operation>
+constexpr void storage_compute_inplace_bitwise(auto const&, auto const& operand2, auto&& result)
+{
+    size_t size2 = min(result.size(), operand2.size());
+
+    IGNORE_VECTOR_DEPENDENCIES
+    for (size_t i = 0; i < size2; ++i) {
+        if constexpr (operation == BitwiseOperation::AND)
+            result[i] &= operand2[i];
+        else if constexpr (operation == BitwiseOperation::OR)
+            result[i] |= operand2[i];
+        else if constexpr (operation == BitwiseOperation::XOR)
+            result[i] ^= operand2[i];
+        else
+            static_assert(((void)operation, false));
+    }
+}
+
+// Requirements for the next two functions:
+//  - shift < result.size() * word_size;
+//  - result.size() == operand.size().
+constexpr void storage_shift_left(auto const& operand, size_t shift, auto&& result)
+{
+    size_t size = operand.size();
+    size_t offset = shift / word_size, remainder = shift % word_size;
+
+    if (shift % word_size == 0) {
+        for (size_t i = size; i-- > offset;)
+            result[i] = operand[i - offset];
+        for (size_t i = 0; i < offset; ++i)
+            result[i] = 0;
+    } else {
+        for (size_t i = size; --i > offset;)
+            result[i] = (operand[i - offset] << remainder) | (operand[i - offset - 1] >> (word_size - remainder));
+        result[offset] = operand[0] << remainder;
+        for (size_t i = 0; i < offset; ++i)
+            result[i] = 0;
+    }
+}
+
+constexpr void storage_shift_right(auto const& operand, size_t shift, auto&& result)
+{
+    size_t size = operand.size();
+    size_t offset = shift / word_size, remainder = shift % word_size;
+
+    if (shift % word_size == 0) {
+        for (size_t i = 0; i < size - offset; ++i)
+            result[i] = operand[i + offset];
+        for (size_t i = size - offset; i < size; ++i)
+            result[i] = 0;
+    } else {
+        for (size_t i = 0; i < size - offset - 1; ++i)
+            result[i] = (operand[i + offset] >> remainder) | (operand[i + offset + 1] << (word_size - remainder));
+        result[size - offset - 1] = operand[size - 1] >> remainder;
+        for (size_t i = size - offset; i < size; ++i)
+            result[i] = 0;
+    }
+}
+
+// Requirements:
+//  - result.size() >= max(operand1.size(), operand2.size()) (not a real constraint but overflow
+//    detection will not work otherwise).
+//
+// Return value:
+// Let r be the return value of the function and a, b, c -- the integer values stored in `operand1`,
+// `operand2` and `result`, respectively. Then,
+//     a + b * (-1) ** subtract = c + r * 2 ** (result.size() * word_size).
+// In particular, r equals 0 iff no overflow has happened.
```

Comment:
"iff' means "if and only if", so it is not a typo :) (https://en.wiktionary.org/wiki/iff)

If the numbers are unsigned, then `r` is just an overflow. In signed case, however, `r`, sign-extended to NativeWord, is the overflowed next word of a result (and it can also be -1 in this case).

I can't return `result` in these functions because it's not trivial how to allocate it. Moreover, creation might also fail (for "dynamic" big integers) and I really don't want to add error handling here.

Reply:
> "iff' means "if and only if", so it is not a typo :) (https://en.wiktionary.org/wiki/iff)

Never saw that until now

> I can't return result in these functions because it's not trivial how to allocate it. Moreover, creation might also fail (for "dynamic" big integers) and I really don't want to add error handling here.

Sadly a side-effect of this being *very* generic, but I see the point